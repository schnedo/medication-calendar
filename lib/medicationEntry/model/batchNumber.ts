export default interface BatchNumber {
  readonly number: number;
}

const prefix = "Ch.-B.";
export function format(batchNumber: BatchNumber): string {
  return `${prefix} ${batchNumber.number}`;
}
