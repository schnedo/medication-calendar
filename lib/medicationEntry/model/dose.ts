export default interface Dose {
  readonly amount: number;
}

const unit = "ml";

export function format(dose: Dose): string {
  return `${dose.amount} ${unit}`;
}
