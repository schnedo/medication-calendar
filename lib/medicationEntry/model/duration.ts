export default interface Duration {
  minutes: number;
  hours: number;
}

function formatPart(number: number | undefined) {
  if (number === undefined) {
    return "00";
  }
  const raw = number.toString();
  return raw.length === 1 ? `0${raw}` : raw;
}

export function format(duration: Duration): string {
  return `${formatPart(duration.hours)}:${formatPart(duration.minutes)}`;
}
