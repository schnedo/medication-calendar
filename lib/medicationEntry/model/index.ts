export { default as BatchNumber } from "./batchNumber";
export { default as Dose } from "./dose";
export { default as Duration } from "./Duration";
export { default as Medicament } from "./medicament";
export { default as Medication } from "./medication";
export { default as MedicationEntry } from "./medicationEntry";

export * as Repository from "./repository";
