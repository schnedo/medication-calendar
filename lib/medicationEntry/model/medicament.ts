import BatchNumber from "./batchNumber";

export default interface Medicament {
  readonly name: string;
  readonly batchNumber: BatchNumber;
}
