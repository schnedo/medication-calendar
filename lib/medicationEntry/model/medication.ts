import Dose from "./dose";
import Medicament from "./medicament";
import { v4 } from "uuid";

export default class Medication {
  readonly id: string = v4();

  constructor(readonly dose: Dose, readonly medicament: Medicament) {}
}
