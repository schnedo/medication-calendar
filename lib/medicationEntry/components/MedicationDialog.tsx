import {
  ChangeEventHandler,
  Dispatch,
  FocusEventHandler,
  PointerEventHandler,
  ReactElement,
  SetStateAction,
  useEffect,
  useRef,
  useState,
} from "react";
import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  FormGroup,
  TextField,
} from "@material-ui/core";
import { ModalProps } from "@material-ui/core/Modal";
import { Medication } from "../model";
import { v4 } from "uuid";

function checkValidity<T>(
  value: T,
  validitySetter: Dispatch<SetStateAction<boolean>>,
): boolean {
  const valid = !!value;
  validitySetter(valid);
  return valid;
}

export interface MedicationDialogProps {
  open: boolean;
  medication?: Medication;
  onClose?: ModalProps["onClose"];
  onSubmit?: (medication: Medication) => Promise<void>;
  onAbort?: () => void;
}

// eslint-disable-next-line sonarjs/cognitive-complexity
export default function MedicationDialog({
  open,
  medication,
  onClose,
  onSubmit,
  onAbort,
}: MedicationDialogProps): ReactElement {
  const idRef = useRef(medication?.id ?? v4());

  const [doseAmount, setDoseAmount] = useState(
    medication?.dose.amount.toString() ?? "",
  );
  const [doseValid, setDoseValid] = useState(true);
  const handleDoseChange: ChangeEventHandler<HTMLInputElement> = ({
    target: { value },
  }) => setDoseAmount(value);
  const handleDoseBlur: FocusEventHandler<HTMLInputElement> = ({
    target: { value },
  }) => checkValidity(value, setDoseValid);
  const handleDoseFocus: FocusEventHandler<HTMLInputElement> = () =>
    setDoseValid(true);

  const [name, setName] = useState(medication?.medicament.name ?? "");
  const [nameValid, setNameValid] = useState(true);
  const handleNameChange: ChangeEventHandler<HTMLInputElement> = ({
    target: { value },
  }) => setName(value);
  const handleNameBlur: FocusEventHandler<HTMLInputElement> = ({
    target: { value },
  }) => checkValidity(value, setNameValid);
  const handleNameFocus: FocusEventHandler<HTMLInputElement> = () =>
    setNameValid(true);

  const [batchNumber, setBatchNumber] = useState(
    medication?.medicament.batchNumber.number.toString() ?? "",
  );
  const [batchNumberValid, setBatchNumberValid] = useState(true);
  const handleBatchNumberChange: ChangeEventHandler<HTMLInputElement> = ({
    target: { value },
  }) => setBatchNumber(value);
  const handleBatchNumberBlur: FocusEventHandler<HTMLInputElement> = ({
    target: { value },
  }) => checkValidity(value, setBatchNumberValid);
  const handleBatchNumberFocus: FocusEventHandler<HTMLInputElement> = () =>
    setBatchNumberValid(true);

  useEffect(() => {
    if (open) {
      const resetState = () => {
        setDoseAmount(medication?.dose.amount.toString() ?? "");
        setDoseValid(true);
        setName(medication?.medicament.name ?? "");
        setNameValid(true);
        setBatchNumber(
          medication?.medicament.batchNumber.number.toString() ?? "",
        );
        setBatchNumberValid(true);
      };
      resetState();
    }
  }, [open, medication]);

  const handleClose: typeof onClose = (event, reason) => {
    onClose && onClose(event, reason);
  };
  const handleAbort = () => {
    onAbort && onAbort();
  };

  const handleSubmit:
    | PointerEventHandler<HTMLButtonElement>
    | undefined = onSubmit
    ? async () => {
        const doseValidity = checkValidity(doseAmount, setDoseValid);
        const nameValidity = checkValidity(name, setNameValid);
        const batchNumberValidity = checkValidity(
          batchNumber,
          setBatchNumberValid,
        );
        if (doseValidity && nameValidity && batchNumberValidity) {
          await onSubmit({
            id: idRef.current,
            medicament: {
              name,
              batchNumber: {
                number: parseInt(batchNumber),
              },
            },
            dose: {
              amount: parseInt(doseAmount),
            },
          });
        }
      }
    : undefined;

  return (
    <Dialog open={open} onClose={handleClose}>
      <DialogTitle>Neue Medikation</DialogTitle>
      <DialogContent>
        <FormGroup>
          <TextField
            id={"doseInput"}
            label={"Dosis"}
            type={"number"}
            value={doseAmount}
            onChange={handleDoseChange}
            onBlur={handleDoseBlur}
            onFocus={handleDoseFocus}
            error={!doseValid}
            helperText={!doseValid && "Bitte Dosis eingeben"}
          />
          <TextField
            id={"nameInput"}
            label={"Name"}
            value={name}
            onChange={handleNameChange}
            onBlur={handleNameBlur}
            onFocus={handleNameFocus}
            error={!nameValid}
            helperText={!nameValid && "Bitte Name eingeben"}
          />
          <TextField
            id={"batchNumberInput"}
            label={"Chargenbezeichnung"}
            type={"number"}
            value={batchNumber}
            onChange={handleBatchNumberChange}
            onBlur={handleBatchNumberBlur}
            onFocus={handleBatchNumberFocus}
            error={!batchNumberValid}
            helperText={
              !batchNumberValid && "Bitte Chargenbezeichnung eingeben"
            }
          />
        </FormGroup>
      </DialogContent>
      <DialogActions>
        <Button color={"primary"} onClick={handleAbort}>
          Abbrechen
        </Button>
        <Button color={"primary"} onClick={handleSubmit}>
          Speichern
        </Button>
      </DialogActions>
    </Dialog>
  );
}
