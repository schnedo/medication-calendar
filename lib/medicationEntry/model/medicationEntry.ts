import Medication from "./medication";
import { BodyMass } from "../../contact/model";
import Duration from "./duration";
import { v4 } from "uuid";

export default class MedicationEntry {
  readonly id: string = v4();

  constructor(
    readonly date: Date,
    readonly medications: Medication[],
    readonly duration: Duration,
    readonly comments: string,
    readonly bodyMass: BodyMass,
  ) {}
}
