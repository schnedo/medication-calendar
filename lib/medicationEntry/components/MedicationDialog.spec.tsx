import { render, waitFor } from "@testing-library/react";
import MedicationDialog from "./MedicationDialog";
import userEvent from "@testing-library/user-event";
import { Medication } from "../model";
import { mock } from "@userlike/joke";

const { v4 } = mock(import("uuid"));

describe("MedicationDialog", () => {
  beforeEach(() => {
    v4.mockReturnValue("uuid");
  });

  it("should render correctly", async () => {
    expect.hasAssertions();

    const { container } = render(<MedicationDialog open={true} />);

    expect(container).toMatchSnapshot();
  });

  it("should call onAbort when pressing abort", async () => {
    expect.hasAssertions();

    const onAbort = jest.fn();
    const { getByText } = render(
      <MedicationDialog open={true} onAbort={onAbort} />,
    );

    userEvent.click(getByText("Abbrechen"));
    expect(onAbort).toHaveBeenCalledTimes(1);
  });

  it("should call onSubmit when submitting", async () => {
    expect.hasAssertions();

    const onSubmit = jest.fn();
    const { getByText, getByLabelText } = render(
      <MedicationDialog open={true} onSubmit={onSubmit} />,
    );

    await userEvent.type(getByLabelText("Dosis"), "200");
    await userEvent.type(getByLabelText("Name"), "name");
    await userEvent.type(getByLabelText("Chargenbezeichnung"), "12345");
    userEvent.click(getByText("Speichern"));

    await waitFor(() => {
      expect(onSubmit).toHaveBeenCalledTimes(1);
    });
    const expected: Medication = {
      dose: {
        amount: 200,
      },
      id: "uuid",
      medicament: {
        name: "name",
        batchNumber: {
          number: 12345,
        },
      },
    };
    expect(onSubmit).toHaveBeenLastCalledWith(expected);
  });

  it("should reset fields when dialog gets closed", async () => {
    expect.hasAssertions();

    const { getByLabelText, rerender } = render(
      <MedicationDialog open={true} />,
    );

    await userEvent.type(getByLabelText("Dosis"), "123");
    await userEvent.type(getByLabelText("Name"), "name");
    await userEvent.type(getByLabelText("Chargenbezeichnung"), "12345");

    rerender(<MedicationDialog open={false} />);
    rerender(<MedicationDialog open={true} />);

    expect(getByLabelText("Dosis")).toHaveDisplayValue("");
    expect(getByLabelText("Name")).toHaveDisplayValue("");
    expect(getByLabelText("Chargenbezeichnung")).toHaveDisplayValue("");
  });

  it("should set fields from provided medication", async () => {
    expect.hasAssertions();

    const medication1: Medication = {
      dose: {
        amount: 200,
      },
      id: "1",
      medicament: {
        name: "name1",
        batchNumber: {
          number: 12345,
        },
      },
    };
    const medication2: Medication = {
      dose: {
        amount: 20,
      },
      id: "2",
      medicament: {
        name: "name2",
        batchNumber: {
          number: 54321,
        },
      },
    };

    const { rerender, getByLabelText } = render(
      <MedicationDialog open={true} medication={medication1} />,
    );

    expect(getByLabelText("Dosis")).toHaveDisplayValue(
      medication1.dose.amount.toString(),
    );
    expect(getByLabelText("Name")).toHaveDisplayValue(
      medication1.medicament.name,
    );
    expect(getByLabelText("Chargenbezeichnung")).toHaveDisplayValue(
      medication1.medicament.batchNumber.number.toString(),
    );

    rerender(<MedicationDialog open={false} />);
    rerender(<MedicationDialog open={true} medication={medication2} />);

    expect(getByLabelText("Dosis")).toHaveDisplayValue(
      medication2.dose.amount.toString(),
    );
    expect(getByLabelText("Name")).toHaveDisplayValue(
      medication2.medicament.name,
    );
    expect(getByLabelText("Chargenbezeichnung")).toHaveDisplayValue(
      medication2.medicament.batchNumber.number.toString(),
    );
  });
});
