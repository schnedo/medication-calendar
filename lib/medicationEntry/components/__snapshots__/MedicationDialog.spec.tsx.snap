// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MedicationDialog should render correctly 1`] = `
<div
  aria-hidden="true"
/>
`;
